### PR TITLE
automl fit with starting points

### PR DESCRIPTION
**Pull request by qingyun-wu**
- Allow fitting with configuration starting points of the estimators
- Better usage of points_to_evaluate in CFO

### CODE REVIEW

File: flaml/automl.py [hunk 1]
```diff
@@ -328,6 +334,12 @@ def best_config(self):
         '''A dictionary of the best configuration.'''
         return self._search_states[self._best_estimator].best_config
 
+    @property
+    def estimators_best_config(self):
+        '''A dictionary of all estimators' best configuration.'''
+        return {e: e_search_state.best_config for (e, e_search_state) in
```
Comment (sonichi) [inline flaml/automl.py:340]:
```suggestion
        return {e: e_search_state.best_config for e, e_search_state in

```

File: flaml/automl.py [hunk 2]
```diff
@@ -873,11 +886,20 @@ def custom_metric(
...
-                validation data
+                validation data.
             groups: None or an array-like of shape (n,) | Group labels for the
                 samples used while splitting the dataset into train/valid set
             verbose: int, default=1 | Controls the verbosity, higher means more
-                messages
+                messages.
+            retrain_full: A bool to specify whether to retrain on the full dataset.
+            split_type: A string to specify the type of the split method
+            learner_selector: a string specifyingthe learner selector
```
Comment (sonichi) [inline flaml/automl.py:896]: not intended for users to set

File: flaml/automl.py [hunk 3]
```diff
@@ -873,11 +886,20 @@ def custom_metric(
...
+            hpo_method: A string to specify the hpo method used.
```
Comment (sonichi) [inline flaml/automl.py:897]: 'grid' is not intended for users

File: flaml/automl.py [hunk 4]
```diff
@@ -873,11 +886,20 @@ def custom_metric(
...
             y_val: None or a numpy array or a pandas series of validation labels
             sample_weight_val: None or a numpy array of the sample weight of
-                validation data
+                validation data.
             groups: None or an array-like of shape (n,) | Group labels for the
                 samples used while splitting the dataset into train/valid set
             verbose: int, default=1 | Controls the verbosity, higher means more
-                messages
+                messages.
+            retrain_full: A bool to specify whether to retrain on the full dataset.
```
Comment (sonichi) [inline flaml/automl.py:894]: not intended for users to set

File: test/test_automl.py [hunk 5]
```diff
@@ -476,6 +476,53 @@ def test_regression_xgboost(self):
         print(automl_experiment.best_loss)
         print(automl_experiment.best_config_train_time)
 
+    def test_fit_w_starting_point(self, as_frame=True):
+        automl_experiment = AutoML()
+        automl_settings = {
+            "time_budget": 20,
```
Comment (sonichi) [inline test/test_automl.py:482]:
```suggestion
            "time_budget": 3,
```

File: test/test_automl.py [hunk 6]
```diff
@@ -476,6 +476,53 @@ def test_regression_xgboost(self):
...
+        automl_val_accuracy = 1.0 - automl_experiment.best_loss
+        print('Best ML leaner:', automl_experiment.best_estimator)
+        print('Best hyperparmeter config:', automl_experiment.best_config)
+        print('Best accuracy on validation data: {0:.4g}'.format(automl_val_accuracy))
+        print('Training duration of best run: {0:.4g} s'.format(automl_experiment.best_config_train_time))
+
+        starting_points = automl_experiment.estimators_best_config
+        print('starting_points', starting_points)
+        automl_settings_resume = {
+            "time_budget": 10,
```
Comment (sonichi) [inline test/test_automl.py:506]:
```suggestion
            "time_budget": 2,
```

File: flaml/automl.py [hunk 7]
```diff
@@ -328,6 +334,12 @@ def best_config(self):
         '''A dictionary of the best configuration.'''
         return self._search_states[self._best_estimator].best_config
 
+    @property
+    def estimators_best_config(self):
```
Comment (sonichi) [inline flaml/automl.py:338]:
```suggestion
    def best_config_per_estimator(self):

```

File: flaml/automl.py [hunk 8]
```diff
@@ -67,8 +67,14 @@ def __init__(self, learner_class, data_size, task):
                     'low_cost_init_value']
             if 'cat_hp_cost' in space:
                 self.cat_hp_cost[name] = space['cat_hp_cost']
+            # if a starting point is provided, set the init config to be
+            # the starting point provided
+            if starting_point is not None and name in starting_point.keys() and \
```
Comment (sonichi) [inline flaml/automl.py:72]:
```suggestion
            if starting_point is not None and starting_point.get(name) is not None:

```

File: flaml/automl.py [hunk 9]
```diff
@@ -949,6 +966,7 @@ def custom_metric(
             self._search_states[estimator_name] = SearchState(
                 learner_class=estimator_class,
                 data_size=self._state.data_size, task=self._state.task,
+                starting_point=starting_points[estimator_name] if estimator_name in starting_points else {},
```
Comment (sonichi) [inline flaml/automl.py:969]:
```suggestion
                starting_point=starting_points.get(estimator_name),

```

**Review by sonichi — CHANGES_REQUESTED** [review]
{} will be used as init_point, which is not what you wanted.
So we should pass None when start_point is not given, instead of {}.

File: flaml/searcher/blendsearch.py [hunk 10]
```diff
@@ -624,9 +687,11 @@ def suggest(self, trial_id: str) -> Optional[Dict]:
         # Number of threads is 1 or 2. Thread 0 is a vacuous thread
         assert len(self._search_thread_pool) < 3, len(self._search_thread_pool)
         if len(self._search_thread_pool) < 2:
-            # When a local converges, the number of threads is 1
+            # When a local thread converges, the number of threads is 1
             # Need to restart
             self._init_used = False
+            if self._candidate_start_points is not None and self._points_to_evaluate:
```
Comment (qingyun-wu) [inline flaml/searcher/blendsearch.py:693]: Assuming self._points_to_evaluate will be suggested first as long as they exist? Need to add comments or move this to suggest.

File: flaml/searcher/blendsearch.py [hunk 11]
```diff
@@ -637,4 +702,31 @@ def _select_thread(self) -> Tuple:
...
+            return True
+
+    def on_trial_complete(self, trial_id: str, result: Optional[Dict] = None,
+                          error: bool = False):
+        super().on_trial_complete(trial_id, result, error)
+        if self._candidate_start_points \
+           and trial_id in self._candidate_start_points:
+            # the trial is a candidate start point
+            if len(self._search_thread_pool) == 2:
+                # ls thread created from this candidate start point
```
Comment (qingyun-wu) [inline flaml/searcher/blendsearch.py:726]: How can we tell this ls thread is created from this candidate start point?

File: flaml/searcher/blendsearch.py [hunk 12]
```diff
@@ -267,25 +289,18 @@ def on_trial_complete(self, trial_id: str, result: Optional[Dict] = None,
             else:  # add to result cache
                 self._result[self._ls.config_signature(config)] = result
                 # update target metric if improved
-                objective = result[
-                    self._metric + self.lagrange] if self._metric_constraints \
-                    else result[self._metric]
+                objective = result[self._ls.metric]
                 if (objective - self._metric_target) * self._ls.metric_op < 0:
                     self._metric_target = objective
                 if not thread_id and metric_constraint_satisfied \
```
Comment (qingyun-wu) [inline flaml/searcher/blendsearch.py:295]: can we change 'not thread_id' to 'thread_id==0'?

File: flaml/searcher/blendsearch.py [hunk 13]
```diff
@@ -637,4 +702,31 @@ def _select_thread(self) -> Tuple:
...
+
+    def on_trial_complete(self, trial_id: str, result: Optional[Dict] = None,
+                          error: bool = False):
+        super().on_trial_complete(trial_id, result, error)
+        if self._candidate_start_points \
+           and trial_id in self._candidate_start_points:
+            # the trial is a candidate start point
+            if len(self._search_thread_pool) == 2:
+                # ls thread created from this candidate start point
+                del self._candidate_start_points[trial_id]
```
Comment (qingyun-wu) [inline flaml/searcher/blendsearch.py:727]: Can you move this line to the place where ls thread is created from the candidate start point, like after line 302.

File: flaml/searcher/blendsearch.py [hunk 14]
```diff
@@ -267,25 +289,18 @@ def on_trial_complete(self, trial_id: str, result: Optional[Dict] = None,
...
-                        self.cost_attr
-                    )
                     thread_id = self._thread_count
-                    self._thread_count += 1
-                    self._update_admissible_region(
-                        config, self._ls_bound_min, self._ls_bound_max)
+                    self._started_from_given = self._candidate_start_points \
+                        and trial_id in self._candidate_start_points
+                    if not self._started_from_given:
+                        self._started_from_low_cost = True
```
Comment (qingyun-wu) [inline flaml/searcher/blendsearch.py:302]: why don't we remove the trial_id from self._candidate_start_points here if a ls thread is created?

File: flaml/searcher/blendsearch.py [hunk 15]
```diff
@@ -132,6 +132,13 @@ def __init__(self,
                 self._gs = GlobalSearch(space=space, metric=metric, mode=mode)
         else:
             self._gs = None
+        if getattr(self, '__name__', None) == 'CFO' and points_to_evaluate and len(
+           points_to_evaluate) > 1:
+            # use the best config in points_to_evaluate as the start point
+            self._candidate_start_points = {}
+            self._started_from_low_cost = not low_cost_partial_config
+        else:
+            self._candidate_start_points = None
```
[the remark this answers is not in the document]
Comment (sonichi) [reply flaml/searcher/blendsearch.py:141]: Right. It's not applicable when self._candidate_start_points is None.

File: flaml/searcher/blendsearch.py (hunk 12 above)
Comment (sonichi) [reply flaml/searcher/blendsearch.py:295]: We can't. thread_id could be None.

**Review by sonichi — APPROVED** [review]

**Pull request merged**